libgalois/include/katana/Range.h

```diff
@@ -32,6 +32,66 @@
 
 namespace katana {
 
+/**
+ * Finds the midpoint of a range.  The first half is always be bigger than
+ * the second half if the range has an odd length.
+ */
+template <typename Iterator>
+Iterator
+split_range(Iterator b, Iterator e) {
+  std::advance(b, (std::distance(b, e) + 1) / 2);
+  return b;
+}
+
+/**
+ * Returns a continuous block from the range based on the number of
+ * divisions and the id of the block requested
+ */
+template <typename Iterator>
+std::enable_if_t<
+    !std::is_integral<Iterator>::value, std::pair<Iterator, Iterator>>
+block_range(Iterator begin, Iterator end, unsigned id, unsigned num) {
+  if (num == 0 || id >= num) {
+    return std::make_pair(end, end);
+  }
+
+  size_t dist = std::distance(begin, end);
+
+  size_t block_size = dist / num;
+  size_t remaining = dist % num;
+
+  size_t b =
+      id < remaining ? id * (block_size + 1) : id * block_size + remaining;
+  size_t e = b + block_size + (id < remaining ? 1 : 0);
```

Comment:
can replace `(id < remaining ? 1 : 0)` with `(id < remaining)`

Reply:
I think the current code is more clear to read. The compiler will optimize it anyway.

Reply:
@roshandathathri : The issue @ddn0 is trying to tackle is of imbalanced assignment of work, which will only be mitigated by work-stealing. 
For example: let's say we have 1000 elements and 30 threads.
Previously each thread would get `ceiling(1000/30) == 34, except last thread would get 1000 - 29*34 = 16 elements`
Now, @ddn0 is deciding that some threads get 33 and some get 34 based on their id.

@ddn0 : The code is certainly harder to read, and I'm not able to follow all of it. I had to do some pencil on paper for the example above :). So, I suggest adding comments . For example, I don't understand the micro-optimizations.

One style that I personally like is making the parameters const values, and computing new values to return. It improves readability IMO.

Reply:
Amber, I understand the code. My comment was a response to Rob's comment, not to the code itself. 

Also, we had faced this issue of imbalance before (when dealing blocks to distributed machines). We couldn't steal work then. But we were not using this block distribution. We were using edge-balanced node-blocking and we wrote that such that it avoiding this imbalance.